app/models/user.rb

```diff
@@ -25,6 +26,47 @@ class User < ApplicationRecord
     password_salt&.last(10)
   end
 
+  generates_token_for :email_confirmation, expires_in: 1.day do
+    unconfirmed_email
+  end
+
+  def pending_email_change?
+    unconfirmed_email.present?
+  end
+
+  def initiate_email_change(new_email)
+    return false if new_email == email
+    return false if new_email == unconfirmed_email
+
+    if Rails.application.config.app_mode.self_hosted? && !Setting.require_email_confirmation
+      update(email: new_email)
+    else
+      if update(
+        unconfirmed_email: new_email,
+        email_confirmation_sent_at: Time.current
+      )
+        EmailConfirmationMailer.with(user: self).confirmation_email.deliver_later
+        true
+      else
+        false
+      end
+    end
+  end
+
+  def confirm_email_change(token)
+    return false if email_confirmation_sent_at < 1.day.ago
+
+    if update(
+      email: unconfirmed_email,
+      unconfirmed_email: nil,
+      email_confirmation_sent_at: nil
+    )
+      true
+    else
+      false
+    end
+  end
```

Comment:
I think we can remove this method and handle directly in the controller.  Then the controller would just be:

```rb
def confirm
  # Returns nil if the token is invalid OR expired
  @user = User.find_by_token_for(:email_confirmation, params[:token])

  if @user&.update(email: @user.unconfirmed_email, unconfirmed_email: nil)
    redirect_to root_path, notice: t(".success_login")
  else
    redirect_to root_path, alert: t(".invalid_token")
  end
end
```